File: backend/excel_to_json.py
import pandas as pd

# Read the Excel file
df = pd.read_excel("./data/cutoffs.xlsx")

# Print original column names for debugging
print("Original columns in Excel file:")
print(df.columns.tolist())

# Clean column names
df.columns = df.columns.str.replace('\n', '').str.strip().str.upper().str.replace(" ", "")

# Print cleaned column names
print("\nCleaned column names:")
print(df.columns.tolist())

# Map Excel column names to our required names
column_mapping = {
    'OC_BOYS': 'OCBOYS',
    'OC_GIRLS': 'OCGIRLS',
    'SC_BOYS': 'SCBOYS',
    'SC_GIRLS': 'SCGIRLS',
    'ST_BOYS': 'STBOYS',
    'ST_GIRLS': 'STGIRLS',
    'BCA_BOYS': 'BCABOYS',
    'BCA_GIRLS': 'BCAGIRLS',
    'BCB_BOYS': 'BCBBOYS',
    'BCB_GIRLS': 'BCBGIRLS',
    'BCC_BOYS': 'BCCBOYS',
    'BCC_GIRLS': 'BCCGIRLS',
    'BCD_BOYS': 'BCDBOYS',
    'BCD_GIRLS': 'BCDGIRLS',
    'BCE_BOYS': 'BCEBOYS',
    'BCE_GIRLS': 'BCEGIRLS',
    'OC_EWS_BOYS': 'OCEWSBOYS',
    'OC_EWS_GIRLS': 'OCEWSGIRLS'
}

# Rename columns
df = df.rename(columns=column_mapping)

# Required columns (based on your schema)
required_cols = [
    'INST_CODE', 'INST_NAME', 'TYPE', 'DIST', 'PLACE', 'COED', 'BRANCH_CODE',
    'OCBOYS', 'OCGIRLS', 'SCBOYS', 'SCGIRLS', 'STBOYS', 'STGIRLS',
    'BCABOYS', 'BCAGIRLS', 'BCBBOYS', 'BCBGIRLS', 'BCCBOYS', 'BCCGIRLS',
    'BCDBOYS', 'BCDGIRLS', 'BCEBOYS', 'BCEGIRLS', 'OCEWSBOYS', 'OCEWSGIRLS',
    'COLLFEE'
]

# Print any missing columns
missing_cols = [col for col in required_cols if col not in df.columns]
if missing_cols:
    print("\nMissing columns:")
    print(missing_cols)

# Fill missing columns with null (None)
for col in required_cols:
    if col not in df.columns:
        df[col] = None

# Select only the required columns
df = df[required_cols]

# Convert rank columns to numeric, replacing any non-numeric values with None
rank_columns = [
    'OCBOYS', 'OCGIRLS', 'SCBOYS', 'SCGIRLS', 'STBOYS', 'STGIRLS',
    'BCABOYS', 'BCAGIRLS', 'BCBBOYS', 'BCBGIRLS', 'BCCBOYS', 'BCCGIRLS',
    'BCDBOYS', 'BCDGIRLS', 'BCEBOYS', 'BCEGIRLS', 'OCEWSBOYS', 'OCEWSGIRLS'
]

for col in rank_columns:
    df[col] = pd.to_numeric(df[col], errors='coerce')

# Save as JSON
df.to_json("./data/college_cutoffs.json", orient="records", indent=2)
print("\n✅ JSON created as './data/college_cutoffs.json'")

File: backend/excel_to_json.js
const XLSX = require('xlsx');
const fs = require('fs');
const path = require('path');

// Function to clean column names
function cleanColumnName(name) {
    return name.replace('\n', '').trim().toUpperCase().replace(' ', '');
}

// Column mapping
const columnMapping = {
    'OC_BOYS': 'OCBOYS',
    'OC_GIRLS': 'OCGIRLS',
    'SC_BOYS': 'SCBOYS',
    'SC_GIRLS': 'SCGIRLS',
    'ST_BOYS': 'STBOYS',
    'ST_GIRLS': 'STGIRLS',
    'BCA_BOYS': 'BCABOYS',
    'BCA_GIRLS': 'BCAGIRLS',
    'BCB_BOYS': 'BCBBOYS',
    'BCB_GIRLS': 'BCBGIRLS',
    'BCC_BOYS': 'BCCBOYS',
    'BCC_GIRLS': 'BCCGIRLS',
    'BCD_BOYS': 'BCDBOYS',
    'BCD_GIRLS': 'BCDGIRLS',
    'BCE_BOYS': 'BCEBOYS',
    'BCE_GIRLS': 'BCEGIRLS',
    'OC_EWS_BOYS': 'OCEWSBOYS',
    'OC_EWS_GIRLS': 'OCEWSGIRLS'
};

// Required columns
const requiredCols = [
    'INST_CODE', 'INST_NAME', 'TYPE', 'DIST', 'PLACE', 'COED', 'BRANCH_CODE',
    'OCBOYS', 'OCGIRLS', 'SCBOYS', 'SCGIRLS', 'STBOYS', 'STGIRLS',
    'BCABOYS', 'BCAGIRLS', 'BCBBOYS', 'BCBGIRLS', 'BCCBOYS', 'BCCGIRLS',
    'BCDBOYS', 'BCDGIRLS', 'BCEBOYS', 'BCEGIRLS', 'OCEWSBOYS', 'OCEWSGIRLS',
    'COLLFEE'
];

// Rank columns for numeric conversion
const rankColumns = [
    'OCBOYS', 'OCGIRLS', 'SCBOYS', 'SCGIRLS', 'STBOYS', 'STGIRLS',
    'BCABOYS', 'BCAGIRLS', 'BCBBOYS', 'BCBGIRLS', 'BCCBOYS', 'BCCGIRLS',
    'BCDBOYS', 'BCDGIRLS', 'BCEBOYS', 'BCEGIRLS', 'OCEWSBOYS', 'OCEWSGIRLS'
];

function convertExcelToJson() {
    try {
        // Read the Excel file
        const workbook = XLSX.readFile('./data/cutoffs.xlsx');
        const sheetName = workbook.SheetNames[0];
        const worksheet = workbook.Sheets[sheetName];
        
        // Convert to JSON
        let data = XLSX.utils.sheet_to_json(worksheet);

        // Print original column names
        console.log("Original columns in Excel file:");
        console.log(Object.keys(data[0]));

        // Clean and map column names
        data = data.map(row => {
            const newRow = {};
            Object.entries(row).forEach(([key, value]) => {
                const cleanKey = cleanColumnName(key);
                const mappedKey = columnMapping[cleanKey] || cleanKey;
                newRow[mappedKey] = value;
            });
            return newRow;
        });

        // Print cleaned column names
        console.log("\nCleaned column names:");
        console.log(Object.keys(data[0]));

        // Check for missing columns
        const missingCols = requiredCols.filter(col => !Object.keys(data[0]).includes(col));
        if (missingCols.length > 0) {
            console.log("\nMissing columns:");
            console.log(missingCols);
        }

        // Fill missing columns with null and ensure only required columns
        data = data.map(row => {
            const newRow = {};
            requiredCols.forEach(col => {
                newRow[col] = row[col] || null;
            });
            return newRow;
        });

        // Convert rank columns to numeric
        data = data.map(row => {
            rankColumns.forEach(col => {
                row[col] = row[col] ? Number(row[col]) : null;
                if (isNaN(row[col])) row[col] = null;
            });
            return row;
        });

        // Save as JSON
        fs.writeFileSync(
            './data/college_cutoffs.json',
            JSON.stringify(data, null, 2)
        );

        console.log("\n✅ JSON created as './data/college_cutoffs.json'");
    } catch (error) {
        console.error("Error converting Excel to JSON:", error);
    }
}

// Run the conversion
convertExcelToJson();
